pageserver/src/tenant/mgr.rs

```diff
@@ -123,9 +198,53 @@ pub async fn init_tenant_mgr(
                         continue;
                     }
 
+                    let tenant_id = match tenant_dir_path
+                        .file_name()
+                        .and_then(OsStr::to_str)
+                        .unwrap_or_default()
+                        .parse::<TenantId>()
+                    {
+                        Ok(id) => id,
+                        Err(_) => {
+                            warn!(
+                                "Invalid tenant path (garbage in our repo directory?): {0}",
+                                tenant_dir_path.display()
+                            );
+                            continue;
+                        }
+                    };
+
+                    let generation = if let Some(generations) = &tenant_generations {
+                        // We have a generation map: treat it as the authority for whether
+                        // this tenant is really attached.
+                        if let Some(gen) = generations.get(&tenant_id) {
+                            Generation::new(*gen)
+                        } else {
+                            info!("Detaching tenant {0}, control plane omitted it in re-attach response", tenant_id);
+                            if let Err(e) = fs::remove_dir_all(&tenant_dir_path).await {
+                                error!(
+                                    "Failed to remove detached tenant directory '{}': {:?}",
+                                    tenant_dir_path.display(),
+                                    e
+                                );
+                            }
```

Comment:
The `remove_dir_all` isn't atomic. (I guess `/detach` has the same problem.)
Non-atomicity is a problem because if we come up again, and for some reason trust the on-disk state again (e.g., control_plane_api is None, or control plane returns the tenant in `/re-atatch` response again), I don't want to think about the implications of a half-deleted tenant directory for `reconcile_with_remote`.

Let's have a preliminary PR that makes detach atomic.
The following sequence should do it:

1. Rename tenant dir to a temporary suffix. (rename is atomic)
2. Fsync parent dir (to make the rename durable)
3. Invariant: if we crash now, pageserver startup is going to resume removal of the dir due to the temorary suffix
4. `fs::remove_dir_all` the temporary dir

WDYT? cc @koivunej

Reply:
This also exists as an existing bug with detach, right? (it also does a remove_dir_all without a tmp step).

In any case I agree that it's sensible to add the change to move directories to a temporary path before starting deleting contents.

Reply:
https://github.com/neondatabase/neon/pull/5182

Reply:
Reviewed it. Let's keep this convo open until that PR is merged.

Reply:
Merged main (incl #5182) in to this PR, and this code path now uses the safe deletion helper.